ingestor/shared/src/test/scala/za/co/absa/hyperdrive/shared/utils/TestClassLoaderUtils.scala

```diff
@@ -0,0 +1,63 @@
+/*
+ *  Copyright 2019 ABSA Group Limited
+ *
+ *   Licensed under the Apache License, Version 2.0 (the "License");
+ *   you may not use this file except in compliance with the License.
+ *   You may obtain a copy of the License at
+ *
+ *       http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *   Unless required by applicable law or agreed to in writing, software
+ *   distributed under the License is distributed on an "AS IS" BASIS,
+ *   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *
+ *   See the License for the specific language governing permissions and
+ *   limitations under the License.
+ */
+
+package za.co.absa.hyperdrive.shared.utils
+
+import org.scalatest.FlatSpec
+
+class TestClassLoaderUtils extends FlatSpec {
+
+  behavior of s"${ClassLoaderUtils.getClass.getSimpleName}.loadSingletonClassOfType"
+
+  it should "return the singleton class for a given trait and fully qualified class name" in {
+    // given
+    val className = TestClassLoaderUtilsSingletonInstanceOf.getClass.getName
+    // when
+    val singleton = ClassLoaderUtils.loadSingletonClassOfType[TestClassLoaderUtilsTestTrait](className)
+    // then
+    assert(singleton.isInstanceOf[TestClassLoaderUtilsTestTrait])
+    assert(singleton.equals(TestClassLoaderUtilsSingletonInstanceOf))
+  }
+
+  it should "throw if the given class does not exist" in {
+    // given
+    val className = "non-existent-class"
+    // when
+    val throwable = intercept[IllegalArgumentException](ClassLoaderUtils.loadSingletonClassOfType[TestClassLoaderUtilsTestTrait](className))
+    // then
+    assert(throwable.getMessage.equals("Class 'non-existent-class' could not be found"))
+  }
+
+  it should "throw if the given class does not implement the given trait" in {
+    // given
+    val className = TestClassLoaderUtilsNotInstanceOf.getClass.getName
+    // when
+    val throwable = intercept[IllegalArgumentException](ClassLoaderUtils.loadSingletonClassOfType[TestClassLoaderUtilsTestTrait](className))
+    // then
+    assert(throwable.getMessage.matches("Class '.*TestClassLoaderUtilsNotInstanceOf.' is not an instance of '.*TestClassLoaderUtilsTestTrait'"))
+  }
+
+  it should "throw if the given class is not a singleton" in {
+    // given
+    val className = (new TestClassLoaderUtilsNoSingleton).getClass.getName
+    // when
+    val throwable = intercept[IllegalArgumentException](ClassLoaderUtils.loadSingletonClassOfType[TestClassLoaderUtilsTestTrait](className))
+    // then
+    assert(throwable.getMessage.contains("TestClassLoaderUtilsNoSingleton' is not a singleton"))
+  }
+
+}
```

Comment:
This is quite awesome!

An interesting test case would be to load a class from an external JAR. I had created [this](https://github.com/AbsaOSS/hyperdrive/blob/develop/ingestor/shared/src/main/scala/za/co/absa/hyperdrive/shared/test/utils/JarTestUtils.scala) utils to allow the creation of a temporary JAR file.

It would be nice to see a JAR being created with a couple of classes and `ClassLoaderUtils` loading them from there.

Reply:
See 859296e. I didn't manage to write a test with a jar created at runtime. To do so, I'd have to add a compiled .class file to the jar, which would be compiled with `mvn install`. But such a class would be available to the classloader anyway, because the class is part of the project. But for the test to be meaningful, the class must only exist in the other jar, but not in the project itself.

Therefore I chose to provide a jar-file as a test resource and a shell script to create that jar-file.

Reply:
Amazing, does the job nicely.